### PR TITLE
feat(replication): Do not auto replicate different master

### LINKED ISSUE #2636: master_id (runid) should protect from accidental master restarts
Currently we use master_id for two purposes: cluster node id generation and as "master id" during the replication.

1. Similarly to Redis, we do not protect slave from accidental data flushes during full sync (see http://antirez.com/news/80 for context)
2. Unlike Redis we currently do not employ partial sync where such protection exists

Not directly related to this issue but important for providing additional context - using master_id as nodeid for cluster management is cumbersome and confusing.

I suggest we implement master_id protection so that replica that had been synced already and reached SSR with the master id A, won't reconnect automatically with master under the same address with master id B.  Specifically, one would need to reissue "replicaof .." command again to bootstrap the replication again.

This behavior change on replica side should be under flag with default to preserve the current behaviour.

### PR DESCRIPTION
**Pull request by chakaz**
Until now, replicas would re-connect and re-replicate a master after the master will restart. This is problematic in case the master loses its data, which will cause the replica to flush all and lose its data as well.

This is a breaking change though, in that whoever controls the replica now has to explicitly issue a `REPLICAOF X Y` in order to re-establish a connection to a new master. This is true even if the master loaded an up to date RDB file.

It's not necessary if the replica lost connection to the master and the master was always alive, and the connection is re-established.

Fixes #2636

### CODE REVIEW

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2127]: await disconnect_clients(c_master, c_replica)

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2093]: Please add a comment what you are checking in this test

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2116]: so you dont need to call master.client() after stop() ?
cause I think we do this in all our test..

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2049]: Maybe rename the func name to is_replicaiton_conn_down

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1406]: So we want to check reconnect with tls here.
Instead of removing this check lets think of another way to make the reconnect work like not using stop to break the connection but use the proxy instead

File: tests/dragonfly/replication_test.py
Comment (chakaz) [reply tests/dragonfly/replication_test.py:2116]: nope, clients automatically reconnect

File: tests/dragonfly/replication_test.py
Comment (chakaz) [reply tests/dragonfly/replication_test.py:1406]: what do you mean by proxy? do we have such a thing?

Comment (adiholden) [reply tests/dragonfly/replication_test.py:1406]: take a look into test_network_disconnect test

Comment (chakaz) [reply tests/dragonfly/replication_test.py:1406]: Nice! Done.

File: src/server/replica.cc
Comment (adiholden) [inline src/server/replica.cc:306]: "This behavior change on replica side should be under flag with default to preserve the current behaviour."

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2110]: This looks like a long line, dont we have a limit ?


    """
Test replica does not connect to master if master restarted   
    step1: create master and replica
    step2: stop master and start again with the same port
    step3: check replica is not replicating the restarted master
    step4: issue new replicaof command 
    step5: check replica replicates master
    """

File: src/server/replica.cc
Comment (chakaz) [reply src/server/replica.cc:306]: I completely missed that :facepalm:

File: src/server/replica.cc
Comment (adiholden) [inline src/server/replica.cc:44]: maybe replica_reconnect_on_master_restart ?

**Comment by adiholden** [comment]
The issue #2636  says "I suggest we implement master_id protection so that replica that had been synced already and reached SSR with the master id A, won't reconnect automatically with master under the same address with master id B"
In this PR we will not reconnect if we got the repl id from master , not necessarily if we reached SSR.
I want to make sure you approve this @romange

File: src/server/replica.cc
Comment (chakaz) [reply src/server/replica.cc]: much better! (done)

**Review by adiholden — DISMISSED** [review]

**Comment by chakaz** [comment]
Thanks @adiholden, waiting for @romange's approval as well (specifically on whether to only abort replication after stable sync unlike what I did)

File: tests/dragonfly/replication_test.py
Comment (adiholden) [inline tests/dragonfly/replication_test.py:35]: move to proxy.py

**Comment by romange** [comment]
Lgtm

**Comment by romange** [comment]
I understand the difference but I think it's manageable, the impact is that for some cases we will have a replica that does not sync. With monitoring we will be alerted of such cases. I am fine waiting to see if this becomes a priblem

**Review by adiholden — APPROVED** [review]

**Pull request merged**
